README.md
GCM Ruby wrapper  [![Build Status](https://secure.travis-ci.org/hifrank/higcm.png?branch=master)](http://travis-ci.org/hifrank/higcm)
===
This is Ruby libray for push message to device via [Google Cloud Messaging for Android](http://developer.android.com/guide/google/gcm/index.html)
# Features
## parallel gcm messaging 
use [Typhoeus](http://typhoeus.github.com/) as http client so it is able to send gcm messges in parallel way.

## handler
parse gcm response with GCM::Handler according to [GCM Response format](http://developer.android.com/guide/google/gcm/gcm.html#response), 
into serveral kinds of events, say, success, fails, retry, unregister, you can only handle events you care about.

# Usage

## send a message
```ruby
sender = HiGCM::Sender.new(your_api_key)
registration_ids = [1, 2, 3]
opts = {
  :collapse_key => "test"
  :data         => { :mesg => "hello GCM" }
}
response = sender.send(registration_ids, opts)
...
```

## send a message with handler
```ruby
# prepare handler for retry and unregister event
handler = HiGCM::Handler.new
handler.do_retry do | retry_ids, opts, response |
  retry_ids.each do | reg_id, retry_after |
    # do retry things
  end
end
#prepare for renew registration_ids 
handler.do_renew_token do | renew_ids, response |
  renew_ids.each do | reg_id, new_reg_id |
    # do renew stuff
  end
end

sender  = HiGCM::Sender.new(your_api_key)
registration_ids = [1, 2, 3]
opts = {
  :collapse_key => "test"
  :data         => { :mesg => "hello GCM" }
}
sender.send(registration_ids, opts, handler)
```

## send a muti-messages in parallel way 
```ruby
sender  = HiGCM::Sender.new(your_api_key)

# queue your messages first 
something.each do | registration_id, name |
  # prepare handler for retry and unregister event
  handler = HiGCM::Handler.new
  handler.do_retry do | retry_ids, opts, response |
    retry_ids.each do | reg_id, retry_after |
      # do retry things
    end
  end
  #prepare for renew registration_ids 
  handler.do_renew_token do | renew_ids, response |
    renew_ids.each do | reg_id, new_reg_id |
      # do renew stuff
    end
  end

  opts = {
    :collapse_key => "test"
    :data         => { :mesg => "hello #{name}" }
  }
  sender.send_async(registration_id, opts, handler)
end
# now fire
sender.send_async_run
```


CHANGELOG
CHANGLOG

** 0.0.4

*** feature

1. convert data payload to String value [issue #3](https://github.com/hifrank/higcm/issues/3) 

*** bug fixing

1. change delay_while_idle to allow while list true, false, [issue #2](https://github.com/hifrank/higcm/issues/2)

** 0.0.3

*** feature

1. return Typhoeus::Response when send instead of Typhoeus::Resquest

*** bug fixing

1. init a handler object if hanlder is not init when sending.


** 0.0.2

*** feature

1. add LICENSE
2. CHANGELOG

*** bug fixing

** 0.0.1
init version


spec/spec_helper.rb
require "rubygems"
require 'json'
require "rspec"

# gem install redgreen for colored test output
begin require "redgreen" unless ENV['TM_CURRENT_LINE']; rescue LoadError; end

path = File.expand_path(File.dirname(__FILE__) + "/../lib/")

$LOAD_PATH.unshift(path) unless $LOAD_PATH.include?(path)

require 'higcm'

RSpec.configure do |config|
end


spec/lib/gcm/sender_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/../../spec_helper')

describe GCM::Sender do

  before(:each) do
    @api_key = 'foo'
    @registration_ids = [1, 2, 3, 4, 5, 6]
    @raw_gcm_response = File.read("#{File.dirname(__FILE__)}/../../fixtures/gcm_response_200.json")
    @retry_after = 10
    @stub_gcm_response = Typhoeus::Response.new(
      :code    => 200,
      :headers => {"Retry-After" => @retry_after},
      :body    =>  @raw_gcm_response,
      :time    => 0.1
    )
    @sender = GCM::Sender.new(@api_key)

  end

  describe "#initialize" do
    it "should raise exception when api_key does not given when init" do
      expect { sender = GCM::Sender.new }.to raise_error
    end

    it "should raise exception when given api_key is nil or empty" do
      expect { sender = GCM::Sender.new ""}.to raise_error
      expect { sender = GCM::Sender.new nil}.to raise_error
    end

    it "should raise exception when given api_key is nil or empty" do
      sender  = GCM::Sender.new(@api_key)
      sender.is_a?(GCM::Sender).should == true
      sender.api_key.should            == @api_key
    end
  end

  describe "#send_async" do
    before(:each) do
      @sender.hydra = Typhoeus::Hydra.new
      @sender.hydra.stub(:post, 'https://android.googleapis.com/gcm/send').and_return(@stub_gcm_response)
    end

    it "should call handler.handle after request is completed" do
      handler = double(GCM::Handler)
      handler.should_receive(:handle).with(@registration_ids, {}, @stub_gcm_response)
      @sender.send_async(@registration_ids, {}, handler)
      @sender.send_async_run
    end

    it "should raise exception if opts[:collapse_key] is not String, empty string is acceptable" do
      expect { @sender.send_async(@registration_ids, {:collapse_key => nil }, GCM::Handler.new) }.to raise_error(GCM::SenderError)
      expect { @sender.send_async(@registration_ids, {:collapse_key => "" }, GCM::Handler.new) }.not_to raise_error(GCM::SenderError)
    end

    it "should raise exception if opts[:data] is not Hash, empty hash is acceptable" do
      expect { @sender.send_async(@registration_ids, {:data => [] }, GCM::Handler.new) }.to raise_error(GCM::SenderError)
      expect { @sender.send_async(@registration_ids, {:data => {} }, GCM::Handler.new) }.not_to raise_error(GCM::SenderError)
    end

    it "should raise exception if opts[:delay_while_idle] && opts[:time_to_live] is not Fixnum" do
      expect { @sender.send_async(@registration_ids, {:delay_while_idle => [] }, GCM::Handler.new) }.to raise_error(GCM::SenderError)
      expect { @sender.send_async(@registration_ids, {:delay_while_idle => 1 }, GCM::Handler.new) }.not_to raise_error(GCM::SenderError)
      expect { @sender.send_async(@registration_ids, {:time_to_live => [] }, GCM::Handler.new) }.to raise_error(GCM::SenderError)
      expect { @sender.send_async(@registration_ids, {:time_to_live => 1 }, GCM::Handler.new) }.not_to raise_error(GCM::SenderError)
    end
  end

  describe "#send_async_run" do
    it "should run Typhoeus::Hydra when send requests" do
      hydra  = double(Typhoeus::Hydra)
      @sender.hydra = hydra
      hydra.should_receive(:run).once
      @sender.send_async_run
    end
  end

end


spec/lib/gcm/handler_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/../../spec_helper')

describe GCM::Handler do
  it "#handle should handle various response in when response code is 500" do
  end

  it "#handle should handle various response when response code is 200" do
    @raw_gcm_response = File.read("#{File.dirname(__FILE__)}/../../fixtures/gcm_response_200.json")
    @retry_after = 10
    @stub_gcm_response = Typhoeus::Response.new(
      :code    => 200,
      :headers => {"Retry-After" => @retry_after},
      :body    =>  @raw_gcm_response,
      :time    => 0.1
    )
    @api_key     = 'foo'
    sender       = GCM::Sender.new(@api_key)
    sender.hydra = Typhoeus::Hydra.hydra
    sender.hydra.stub(:post, 'https://android.googleapis.com/gcm/send').and_return(@stub_gcm_response)

    _fails   = 0
    _success = 0
    _renew   = 0
    _retry   = 0

    _updated_token = { 5 => "32"}

    handler = GCM::Handler.new
    handler.do_success do | succes_ids, response |
      @success_ids      = succes_ids
      @success_response = response   
    end
    handler.do_retry do | retry_ids, opts, response |
      @retry_ids      = retry_ids
      @retry_opts     = opts
      @retry_response = response
    end
    handler.do_fail do | fail_ids, response |
      @fail_ids      = fail_ids
      @fail_response = response
    end
    handler.do_renew_token do | renew_ids, response |
      @renew_ids      = renew_ids
      @renew_response = response
    end

    registration_ids = [1, 2, 3, 4, 5, 6]
    sender.send_async(registration_ids, {}, handler)
    sender.send_async_run

    @fail_ids.should == { 2 => "Unavailable, retry after #{@retry_after}", 3 => "InvalidRegistration", 6 => "NotRegistered" } 
    @fail_response.is_a?(Typhoeus::Response).should == true

    @retry_ids.should  == { 2 => 10 } 
    @retry_opts.should == {} 
    @retry_response.is_a?(Typhoeus::Response).should == true

    @renew_ids.should == { 5 => "32" } 
    @renew_response.is_a?(Typhoeus::Response).should == true

    @success_ids.should == { 1 => "1:0408", 4 => "1:1516", 5 => "1:2342" } 
    @success_response.is_a?(Typhoeus::Response).should == true

  end

end


lib/higcm/sender.rb
require 'typhoeus'

module HiGCM

    class SenderError < StandardError; end

    class Sender
      attr_accessor :api_key, :api_status, :hydra, :requests

      OPTIONAL_OPTIONS = {
        :collapse_key     => String,
        :data             => Hash,
        :delay_while_idle => [true, false],
        :time_to_live     => Fixnum
      }

      def initialize(api_key)
        @api_key  = api_key
        raise SenderError.new("api_key is necessary for #{self.class}") if api_key.nil? || api_key.empty?
      end

      def send(registration_ids, opts={}, handler=nil)
        handler = HiGCM::Handler.new if handler.nil?
        request = send_async(registration_ids, opts, handler)
        send_async_run
        request.handled_response
      end

      #http://developer.android.com/guide/google/gcm/gcm.html#server
      def send_async(registration_ids, opts={}, handler=nil)

        headers = {
          'Content-Type'  => 'application/json',
          'Authorization' => sprintf("key=%s", @api_key)
        }

        body = {
          'registration_ids' => registration_ids,
        }

        #fill up option
        OPTIONAL_OPTIONS.each do | key, type |
          if opts.key?(key)
            if type.is_a?(Array)
              @valid_value = false
              type.each do | v |
                if opts[key] == v
                  @valid_value = true
                  break
                end
              end
              raise SenderError.new("#{key} should be Type #{type}") unless @valid_value
            else
              raise SenderError.new("#{key} should be Type #{type}") unless opts[key].is_a?(type)
            end
            # convert payload data to String for issue #3
            case key
            when :data
              opts[key] = convert_hash(opts[key])
            end
            body[key] = opts[key]
          end
        end

        request = Typhoeus::Request.new(
          'https://android.googleapis.com/gcm/send',
          :headers         => headers,
          :method          => :post,
          :body            => body.to_json,
          :follow_location => true
        )

        @hydra ||= Typhoeus::Hydra.new

        request.on_complete do | response |
          handler.handle(registration_ids, opts, response)
        end

        @hydra.queue(request)

        request
      end

      def send_async_run
        # handle response according to http://developer.android.com/guide/google/gcm/gcm.html#response
        @hydra.run
      end

      def convert_hash(hash)
        hash.each do | k, v |
          if v.is_a?(Hash)
            hash[k] = convert_hash(v)
          else
            if v.respond_to?(:to_s)
              hash[k] = v.to_s
            else
              raise SenderError "data value must respond to to_s function for converting to String"
            end
          end
        end
      end

    end
end


lib/higcm/handler.rb
require 'json'

module HiGCM

  class Handler

    attr_accessor :retry_conditions, :unregister_conditions

    def initialize
      @retry_ids   = {}
      @unreg_ids   = {}
      @fail_ids    = {}
      @success_ids = {}
      @renew_ids   = {}

      @retry_conditions      = [ 'InternalServerError', 'Timout', 'Unavailable' ]
      @unregister_conditions = [ 'NotRegistered', 'InvalidRegistration' ]
    end

    def handle(registration_ids, opts, response)
      begin
        @code = response.code.to_i
        @body = JSON.parse(response.body)
        #Honor Retry-After
        @retry_after = (response.headers['Retry-After'].nil? ? 0 : response.headers['Retry-After'])
      rescue Exception => e
        @code = 99
        @error_message = "unexpected error, response: #{response.body}, exception: #{e.inspect}"
      end

      case @code
      when 200
        #200 Message was processed successfully. The response body will contain more details about the message status, but its format will depend whether the request was JSON or plain text. See Interpreting a success response for more details.
        # handle success case, http://developer.android.com/guide/google/gcm/gcm.html#success
        @body['results'].each_with_index do | rs, index |
          reg_id = registration_ids[index]
          #handle fail
          if rs.key?('error')
            @error_message = rs['error']
            # figure out retry stuff
            if @retry_conditions.include?(rs['error'])
              @retry_ids[reg_id] = @retry_after
              @error_message << ", retry after #{@retry_after}"
            end
            # figure out unregister stuff
            if @unregister_conditions.include?(rs['error'])
              @unreg_ids[reg_id] = rs['error']
            end
            @fail_ids[reg_id] = @error_message
          #handle success
          elsif rs.key?('message_id')
            @success_ids[reg_id] = rs['message_id']
            if rs.key?('registration_id')
              @renew_ids[reg_id] = rs['registration_id']
            end
          else
            #should not jump here
          end
        end
        @do_success.call(@success_ids, response)        if @success_ids.count > 0 && @do_success
        @do_fail.call(@fail_ids, response)              if @fail_ids.count > 0 && @do_fail
        @do_renew_token.call(@renew_ids, response)      if @renew_ids.count > 0 && @do_renew_token
        @do_retry.call(@retry_ids, opts, response)      if @retry_ids.count > 0 && @do_retry
        @do_unregister_token.call(@unreg_ids, response) if @unreg_ids.count > 0 && @do_unregister_token

      #TODO need to check what kinf of response will return
      when 400, 401
        #400 Only applies for JSON requests. Indicates that the request could not be parsed as JSON, or it contained invalid fields (for instance, passing a string where a number was expected). The exact failure reason is described in the response and the problem should be addressed before the request can be retried.
        #401 There was an error authenticating the sender account.
        registration_ids.each do | reg_id |
          if 400 == @code
            error_message = 'request could not be parsed as JSON'
          else
            error_message = 'There was an error authenticating the sender account'
          end
          @fail_ids[reg_id] = error_message
        end
        @do_fail.call(@fail_ids, response) if @do_fail &&  @fail_ids.count > 0
      #TODO need to check what kinf of response will return
      when 500, 503
        #500 There was an internal error in the GCM server while trying to process the request. trouble shooting http://developer.android.com/guide/google/gcm/gcm.html#internal_error
        #503 Indicates that the server is temporarily unavailable (i.e., because of timeouts, etc ). Sender must retry later, honoring any Retry-After header included in the response. Application servers must implement exponential back-off. The GCM server took too long to process the request. Troubleshoot
        registration_ids.each do | reg_id |
          @retry_ids[reg_id] = @retry_after
        end
        @do_retry.call(@retry_ids, opts) if @do_retry && @retry_ids.count > 0
      else
        registration_ids.each do | reg_id |
          @fail_ids[reg_id] = @error_message
        end
        @do_fail.call(@fail_ids, response) if @do_fail && @fail_ids.count > 0
      end
    end

    # success_ids
    def do_success(&block)
      @do_success = block
    end

    # retry_ids, opts, retry_after
    def do_retry(&block)
      @do_retry = block
    end

    def do_fail(&block)
      @do_fail = block
    end

    # renew_ids(hash)
    def do_renew_token(&block)
      @do_renew_token = block
    end

    def do_unregister_token(&block)
      @do_unregister_token = block
    end

  end
end
